File: Data Structures & Algorithms/duplicate-integer/submission-4.go
func sortArray(nums [] int) [] int {
    var bubble int
    sorted  := nums
    for i := 1; i < len(sorted); i++ {
        for j := 0; j < len(sorted) - i; j++ {
            if sorted[j] > sorted[j + 1] {
                bubble = sorted[j]
                sorted[j] = sorted[j + 1]
                sorted[j + 1] = bubble
            }
        }
    }

    return sorted
}

func Merge(left [] int, right [] int) [] int {
    var merged [] int

    if left[0] < right[1] {
        merged = append(left, right...)
    } else {
        merged = append(right, left...)
    }
    return merged
}

func mergeSort(nums []int) [] int {

    if len(nums) < 2 {
        return nums
    }

    mid := len(nums) / 2

    left := nums[:mid]
    right := nums[mid:]

    left = mergeSort(left)
    right = mergeSort(right)

    return Merge(left, right)
}

func hasDuplicate(nums []int) bool {
    // sort and check current and next values
    sorted := sortArray(nums)

    for i := 1; i < len(sorted); i++ {
        if (sorted[i] == sorted[i-1]) {
            return true
        }
    }
    return false
}

File: Data Structures & Algorithms/binary-search/submission-4.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // the final aproach is to use the left and rigth indices
        // [-1,0,3,5,9,12] t = 9
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }

        }
        return -1;
    }
};
